functorch/csrc/BatchRulesBinaryOps.cpp

```diff
@@ -163,6 +175,52 @@ Tensor addr_decomposition(
   return self * beta + outer;
 }
 
+std::tuple<Tensor,optional<int64_t>> cdist_backward_batch_rule(
+    const Tensor& grad, optional<int64_t> grad_bdim,
+    const Tensor& x1, optional<int64_t> x1_bdim,
+    const Tensor& x2, optional<int64_t> x2_bdim,
+    const double p,
+    const Tensor& cdist, optional<int64_t> cdist_bdim) {
+
+  auto x1_ = x1;
+  if (cdist_bdim && !x1_bdim) {
+    // We need to make sure that x1 has batch dim if cdist has one
+    // otherwise, we get
+    // RuntimeError: Function CdistBackward0 returned an invalid gradient at index 1 - got [5]
+    // but expected shape compatible with [4, 5]
+    auto bs = cdist.size(*cdist_bdim);
+    x1_ = ensure_has_bdim(x1, false, bs);
+    x1_ = x1_.contiguous();
+    x1_bdim = 0;
+  }
+
+  // We need to apply the same preprocessing on x1 and x2 as in the forward pass
+  // _binary_pointwise_batch_rule
+  auto x12 = _binary_pointwise_helper(x1_, x1_bdim, x2, x2_bdim);
```

Comment:
Does cdist do type promotion? `_binary_pointwise_helper` handles type promotion; if cdist doesn't do type promotion then this will lead to different behavior than cdist

Reply:
Thanks for a thourough review @zou3519 ! Yes, _binary_pointwise_helper performs a dtype promotion if one arg is a scalar, but cdist can't work with scalar, at least 2d tensors: https://github.com/pytorch/pytorch/blob/8a912014b166c97fcaf2b89e6ff5a466d1a93f7b/aten/src/ATen/native/Distance.cpp#L119-L120

So, using `_binary_pointwise_helper` on scalars we change the type but cdist will complain about dimensions.
If we are using correct shapes and incorrect dtypes, IMO there is no dtype promotion by cdist batch rule and original cdist will complain about non-floating dtype...

Reply:
Good point, thanks for the reasoning!